### ISSUE TITLE
Bugs in our synchronization parameter update implementation.

### ISSUE THREAD

**Issue by gongweibao** [PaddlePaddle/Paddle#7643]
The synchronization mechanism was implemented in [this file](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/recv_op.cc)

After received all gradients from all trainers, a trainer can get parameters from here: https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/recv_op.cc#L166

And a trainer can  do both send grad and get parameter in the range https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/recv_op.cc#L110 to https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/recv_op.cc#L137.

```
for (size_t i = 0; i < param_count * trainer_count; ++i) {
        // blocking get one var from client.
        const detail::MessageWithName &v = rpc_service_->Get();
        auto grad_var_name = v.first;
        ....
        auto it = std::find(grad_list.begin(), grad_list.end(), grad_var_name);
        std::string param_var_name;
        if (it != grad_list.end()) {
          param_var_name = param_list[it - grad_list.begin()];
        } else {
          LOG(ERROR) << "grad have no paired param found!\"" << grad_var_name
                     << "\"";
        }
        VLOG(3) << "recved grad: " << grad_var_name
                << " updating param: " << param_var_name;

        auto *merged_grad = recv_scope.FindVar(grad_var_name);
        ...
        if (trainer_count > 1) {
          // NOTICE: get variable name from a increament count.
          grad_var_name = this->GetGradVarNameForTrainer(grad_var_name);
        }

        auto *var = recv_scope.Var(grad_var_name);
        detail::DeserializeFromMessage(v.second, dev_ctx, var);
      }
```

If a trainer run much faster than others, it can complete sending and get in that range. So it can send the next pass gradient to pserver. Since we generate a variable name by count:https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/recv_op.cc#L81, so the variable name's id may larger than the number of trainers.

Let's see the log, please notice the <span style="color:red">**time sequence**</span>.

It's the fastest trainer's log:
```
2732 I0118 04:36:07.604624 21361 send_op.cc:47]  begin send grad count:14
2733 I0118 04:36:07.604873 21361 grpc_client.cc:47] send conv2d_0.w_0@GRAD end_point:21361 pserver:127.0.0.1:6170
......
2746 I0118 04:36:07.606597 21361 grpc_client.cc:47] send conv2d_1.w_0@GRAD.block4 end_point:21361 pserver:127.0.0.1:6173
# pserver is still receiving gradients.
2747 I0118 04:36:07.608996 21361 grpc_client.cc:87] get conv2d_0.w_0 pserver:127.0.0.1:6170
....
2760 I0118 04:36:07.610491 21361 grpc_client.cc:87] get conv2d_1.w_0.block4 pserver:127.0.0.1:6173
2761 batch_id 91, loss: 0.535763, acc: 0.611304
#It completes send and get of one pass, and send gradient again.
2762 I0118 04:36:07.927748 21361 send_op.cc:47]  begin send grad count:14
# pserver is still receiving gradients last pass.
2763 I0118 04:36:07.928071 21361 grpc_client.cc:47] send conv2d_0.w_0@GRAD end_point:21361 pserver:127.0.0.1:6170
2764 I0118 04:36:07.928234 21361 grpc_client.cc:47] send fc_0.b_0@GRAD end_point:21361 pserver:127.0.0.1:6171
```

The pserver's log:
```
3021 I0118 04:36:07.604545 21282 recv_op.cc:113]  begin recv param_cout:3 fan_in:5
3022 I0118 04:36:07.607034 21505 grpc_server.cc:67] process var_name:conv2d_1.w_0@GRAD.block1 end_point:21361 len:20030
3024 I0118 04:36:07.607174 21505 grpc_server.cc:67] process var_name:fc_0.w_0@GRAD.block2 end_point:21361 len:6427
3026 I0118 04:36:07.607300 21505 grpc_server.cc:67] process var_name:conv2d_0.w_0@GRAD end_point:21361 len:2030
......
# pserver is still receiving gradients.
3046 I0118 04:36:07.930323 21505 grpc_server.cc:67] process var_name:conv2d_1.w_0@GRAD.block1 end_point:21361 len:20030
3048 I0118 04:36:07.930527 21505 grpc_server.cc:67] process var_name:fc_0.w_0@GRAD.block2 end_point:21361 len:6427
3049 I0118 04:36:07.930657 21505 grpc_server.cc:67] process var_name:conv2d_0.w_0@GRAD end_point:21361 len:2030
3052 I0118 04:36:07.932749 21505 grpc_server.cc:67] process var_name:conv2d_1.w_0@GRAD.block1 end_point:21359 len:20030
3053 I0118 04:36:07.932884 21505 grpc_server.cc:67] process var_name:fc_0.w_0@GRAD.block2 end_point:21359 len:6427
3054 I0118 04:36:07.933001 21505 grpc_server.cc:67] process var_name:conv2d_0.w_0@GRAD end_point:21359 len:2030
I0118 04:36:07.936682 21282 recv_op.cc:113]  begin recv param_cout:3 fan_in:5
...
# error log:
E0118 04:36:08.262740 21282 recv_op.cc:140] can not find server side var: conv2d_1.w_0@GRAD.block1.trainer_5
```

**Comment by typhoonzero** [comment]
Trainers cannot enter the next batch until the server has received enough gradients to calculate because the line `const detail::MessageWithName &v = rpc_service_->Get();` is blocking.

**Comment by gongweibao** [comment]
They are three threads:
- process send request
   - request can insert to `rpc_service_.queue_` any time.
- process get request
   - once https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/recv_op.cc#L166 set `done_` flag, the trainer can get until https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/recv_op.cc#L151
- executor.run

**Comment by typhoonzero** [comment]
Well, the reason for this symptom is that we run the GRPC service in a separated thread, the service writes data to a queue, and recv_op read data from the queue. Since we only do barrier in the queue but we have no idea that when the server should go to into a barrier.
